File: Freedom/微信/我/ExpressionViewController/WXExpressionChosenViewController.m
//  TLExpressionChosenViewController.m
//  Freedom
//  Created by Super on 16/4/4.
#import "WXExpressionChosenViewController.h"
#import "WXExpressionSearchViewController.h"
#import "WXSearchController.h"
#import "WXExpressionDetailViewController.h"
#import "WXExpressionHelper.h"
#import "WXPictureCarouselView.h"
#define         EDGE_TOP        10.0f
#define         EDGE_LEFT       15.0f
#define         ROW_SPCAE       10.0f
@interface WXExpressionCell ()
@property (nonatomic, strong) UIImageView *iconImageView;
@property (nonatomic, strong) UILabel *titleLabel;
@property (nonatomic, strong) UILabel *detailLabel;
@property (nonatomic, strong) UIImageView *tagView;
@property (nonatomic, strong) UIButton *downloadButton;
@end
@implementation WXExpressionCell
- (id)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier{
    if (self = [super initWithStyle:style reuseIdentifier:reuseIdentifier]) {
        [self.contentView addSubview:self.iconImageView];
        [self.contentView addSubview:self.titleLabel];
        [self.contentView addSubview:self.detailLabel];
        [self.contentView addSubview:self.tagView];
        [self.contentView addSubview:self.downloadButton];
        
        [self p_addMasonry];
    }
    return self;
}
- (void)setGroup:(TLEmojiGroup *)group{
    _group = group;
    UIImage *image = [UIImage imageNamed:group.groupIconPath];
    if (image) {
        [self.iconImageView setImage:image];
    }else{
        [self.iconImageView sd_setImageWithURL:[NSURL URLWithString:group.groupIconURL]];
    }
    [self.titleLabel setText:group.groupName];
    [self.detailLabel setText:group.groupDetailInfo];
    
    if (group.status == TLEmojiGroupStatusDownloaded) {
        [self.downloadButton setTitle:@"已下载" forState:UIControlStateNormal];
        [self.downloadButton.layer setBorderColor:[UIColor grayColor].CGColor];
        [self.downloadButton setTitleColor:[UIColor grayColor] forState:UIControlStateNormal];
    }else if (group.status == TLEmojiGroupStatusDownloading) {
        [self.downloadButton setTitle:@"下载中" forState:UIControlStateNormal];
        [self.downloadButton.layer setBorderColor:[UIColor greenColor].CGColor];
        [self.downloadButton setTitleColor:[UIColor greenColor] forState:UIControlStateNormal];
    }else{
        [self.downloadButton setTitle:@"下载" forState:UIControlStateNormal];
        [self.downloadButton.layer setBorderColor:[UIColor greenColor].CGColor];
        [self.downloadButton setTitleColor:[UIColor greenColor] forState:UIControlStateNormal];
    }
}
#pragma mark - Event Response
- (void)downloadButtonDown:(UIButton *)sender{
    if (self.group.status == TLEmojiGroupStatusUnDownload) {
        self.group.status = TLEmojiGroupStatusDownloading;
        [self setGroup:self.group];
        if (_delegate && [_delegate respondsToSelector:@selector(expressionCellDownloadButtonDown:)]) {
            [_delegate expressionCellDownloadButtonDown:self.group];
        }
    }
}
#pragma mark - Private Methods
- (void)p_addMasonry{
    [self.iconImageView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.mas_equalTo(self.contentView).mas_offset(EDGE_LEFT);
        make.top.mas_equalTo(self.contentView).mas_offset(EDGE_TOP);
        make.bottom.mas_equalTo(self.contentView).mas_offset(-EDGE_TOP);
        make.width.mas_equalTo(self.iconImageView.mas_height);
    }];
    [self.titleLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.bottom.mas_equalTo(self.iconImageView.mas_centerY).mas_offset(-2.0f);
        make.left.mas_equalTo(self.iconImageView.mas_right).mas_offset(13.0f);
        make.right.mas_lessThanOrEqualTo(self.downloadButton.mas_left).mas_offset(-EDGE_LEFT);
    }];
    [self.detailLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.mas_equalTo(self.iconImageView.mas_centerY).mas_offset(5.0);
        make.left.mas_equalTo(self.titleLabel);
        make.right.mas_lessThanOrEqualTo(self.downloadButton.mas_left).mas_offset(-EDGE_LEFT);
    }];
    [self.tagView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.and.top.mas_equalTo(self.contentView);
    }];
    [self.downloadButton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.right.mas_equalTo(self.contentView).mas_offset(-EDGE_LEFT);
        make.centerY.mas_equalTo(self.contentView);
        make.size.mas_equalTo(CGSizeMake(60, 26));
    }];
}
#pragma mark - Getter
- (UIImageView *)iconImageView{
    if (_iconImageView == nil) {
        _iconImageView = [[UIImageView alloc] init];
        [_iconImageView setBackgroundColor:[UIColor clearColor]];
        [_iconImageView.layer setMasksToBounds:YES];
        [_iconImageView.layer setCornerRadius:5.0f];
    }
    return _iconImageView;
}
- (UILabel *)titleLabel{
    if (_titleLabel == nil) {
        _titleLabel = [[UILabel alloc] init];
        [_titleLabel setFont:[UIFont systemFontOfSize:16.0f]];
    }
    return _titleLabel;
}
- (UILabel *)detailLabel{
    if (_detailLabel == nil) {
        _detailLabel = [[UILabel alloc] init];
        [_detailLabel setFont:[UIFont systemFontOfSize:13.0f]];
        [_detailLabel setTextColor:[UIColor grayColor]];
    }
    return _detailLabel;
}
- (UIImageView *)tagView{
    if (_tagView == nil) {
        _tagView = [[UIImageView alloc] init];
        [_tagView setImage:[UIImage imageNamed:@"icon_corner_new"]];
        [_tagView setHidden:YES];
    }
    return _tagView;
}
- (UIButton *)downloadButton{
    if (_downloadButton == nil) {
        _downloadButton = [[UIButton alloc] init];
        [_downloadButton setTitle:@"下载" forState:UIControlStateNormal];
        [_downloadButton.titleLabel setFont:[UIFont systemFontOfSize:14.0f]];
        [_downloadButton setTitleColor:[UIColor greenColor] forState:UIControlStateNormal];
        [_downloadButton.layer setMasksToBounds:YES];
        [_downloadButton.layer setCornerRadius:3.0f];
        [_downloadButton.layer setBorderWidth:1.0f];
        [_downloadButton.layer setBorderColor:[UIColor greenColor].CGColor];
        [_downloadButton addTarget:self action:@selector(downloadButtonDown:) forControlEvents:UIControlEventTouchUpInside];
    }
    return _downloadButton;
}
@end
@interface WXExpressionBannerCell () <WXPictureCarouselDelegate>
@property (nonatomic, strong) WXPictureCarouselView *picCarouselView;
@end
@implementation WXExpressionBannerCell
- (id)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier{
    if (self = [super initWithStyle:style reuseIdentifier:reuseIdentifier]) {
        [self setBottomLineStyle:TLCellLineStyleNone];
        [self setSelectionStyle:UITableViewCellSelectionStyleNone];
        [self.contentView addSubview:self.picCarouselView];
        
        [self p_addMasonry];
    }
    return self;
}
- (void)setData:(NSArray *)data{
    _data = data;
    [self.picCarouselView setData:data];
}
#pragma mark - 
- (void)pictureCarouselView:(WXPictureCarouselView *)pictureCarouselView didSelectItem:(id<WXPictureCarouselProtocol>)model{
    if (self.delegate && [self.delegate respondsToSelector:@selector(expressionBannerCellDidSelectBanner:)]) {
        [self.delegate expressionBannerCellDidSelectBanner:model];
    }
}
#pragma mark - 
- (void)p_addMasonry{
    [self.picCarouselView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.mas_equalTo(self.contentView);
    }];
}
#pragma mark - 
- (WXPictureCarouselView *)picCarouselView{
    if (_picCarouselView == nil) {
        _picCarouselView = [[WXPictureCarouselView alloc] init];
        [_picCarouselView setDelegate:self];
    }
    return _picCarouselView;
}
@end
@interface WXExpressionChosenViewController () <UISearchBarDelegate>
@property (nonatomic, strong) WXSearchController *searchController;
@property (nonatomic, strong) WXExpressionSearchViewController *searchVC;
@end
@implementation WXExpressionChosenViewController
- (void)viewDidLoad{
    [super viewDidLoad];
    [self setAutomaticallyAdjustsScrollViewInsets:NO];
    [self.tableView setFrame:CGRectMake(0, TopHeight + 20, APPW, APPH - 20 - TopHeight)];
    [self.tableView setBackgroundColor:[UIColor whiteColor]];
    [self.tableView setTableHeaderView:self.searchController.searchBar];
    
    MJRefreshAutoNormalFooter *footer = [MJRefreshAutoNormalFooter footerWithRefreshingTarget:self refreshingAction:@selector(loadMoreData)];
    [footer setTitle:@"正在加载..." forState:MJRefreshStateRefreshing];
    [footer setTitle:@"" forState:MJRefreshStateNoMoreData];
    [self.tableView setMj_footer:footer];
    
    [self registerCellsForTableView:self.tableView];
    [self loadDataWithLoadingView:YES];
}
- (void)viewDidDisappear:(BOOL)animated{
    [super viewDidDisappear:animated];
    [SVProgressHUD dismiss];
}
#pragma mark - 
- (WXExpressionHelper *)proxy{
    if (_proxy == nil) {
        _proxy = [WXExpressionHelper sharedHelper];
    }
    return _proxy;
}
- (WXSearchController *)searchController{
    if (_searchController == nil) {
        _searchController = [[WXSearchController alloc] initWithSearchResultsController:self.searchVC];
        [_searchController setSearchResultsUpdater:self.searchVC];
        [_searchController.searchBar setPlaceholder:@"搜索表情"];
        [_searchController.searchBar setDelegate:self.searchVC];
    }
    return _searchController;
}
- (WXExpressionSearchViewController *)searchVC{
    if (_searchVC == nil) {
        _searchVC = [[WXExpressionSearchViewController alloc] init];
    }
    return _searchVC;
}
- (void)loadDataWithLoadingView:(BOOL)showLoadingView{
    if (showLoadingView) {
        [SVProgressHUD show];
    }
    kPageIndex = 1;
    __weak typeof(self) weakSelf = self;
    [self.proxy requestExpressionChosenListByPageIndex:kPageIndex success:^(id data) {
        [SVProgressHUD dismiss];
        kPageIndex ++;
        weakSelf.data = [[NSMutableArray alloc] init];
        for (TLEmojiGroup *group in data) {     // 优先使用本地表情
            TLEmojiGroup *localEmojiGroup = [[WXExpressionHelper sharedHelper] emojiGroupByID:group.groupID];
            if (localEmojiGroup) {
                [self.data addObject:localEmojiGroup];
            }else{
                [self.data addObject:group];
            }
        }
        [weakSelf.tableView reloadData];
    } failure:^(NSString *error) {
        [SVProgressHUD dismiss];
    }];
    
    [self.proxy requestExpressionChosenBannerSuccess:^(id data) {
        self.bannerData = data;
        [self.tableView reloadData];
    } failure:^(NSString *error) {
        
    }];
}
- (void)loadMoreData{
    __weak typeof(self) weakSelf = self;
    [self.proxy requestExpressionChosenListByPageIndex:kPageIndex success:^(NSMutableArray *data) {
        [SVProgressHUD dismiss];
        if (data.count == 0) {
            [self.tableView.mj_footer endRefreshingWithNoMoreData];
        }else{
            [self.tableView.mj_footer endRefreshing];
            kPageIndex ++;
            for (TLEmojiGroup *group in data) {     // 优先使用本地表情
                TLEmojiGroup *localEmojiGroup = [[WXExpressionHelper sharedHelper] emojiGroupByID:group.groupID];
                if (localEmojiGroup) {
                    [self.data addObject:localEmojiGroup];
                }
                else {
                    [self.data addObject:group];
                }
            }
            [weakSelf.tableView reloadData];
        }
    } failure:^(NSString *error) {
        [self.tableView.mj_footer endRefreshingWithNoMoreData];
        [SVProgressHUD dismiss];
    }];
}
- (void)registerCellsForTableView:(UITableView *)tableView{
    [tableView registerClass:[WXExpressionBannerCell class] forCellReuseIdentifier:@"TLExpressionBannerCell"];
    [tableView registerClass:[WXExpressionCell class] forCellReuseIdentifier:@"TLExpressionCell"];
}
#pragma mark - 
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView{
    return self.bannerData.count > 0 ? 2 : 1;
}
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    if (section == 0) {
        return self.bannerData.count > 0 ? 1 : self.data.count;
    }else if (section == 1) {
        return self.data.count;
    }
    return 0;
}
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    if (indexPath.section == 0 && self.bannerData.count > 0) {
        WXExpressionBannerCell *bannerCell = [tableView dequeueReusableCellWithIdentifier:@"TLExpressionBannerCell"];
        [bannerCell setData:self.bannerData];
        [bannerCell setDelegate:self];
        return bannerCell;
    }
    WXExpressionCell *cell = [tableView dequeueReusableCellWithIdentifier:@"TLExpressionCell"];
    TLEmojiGroup *group = self.data[indexPath.row];
    [cell setGroup:group];
    [cell setDelegate:self];
    return cell;
}
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    if ((indexPath.section == 0 && self.bannerData.count == 0) || indexPath.section == 1) {
        TLEmojiGroup *group = [self.data objectAtIndex:indexPath.row];
        WXExpressionDetailViewController *detailVC = [[WXExpressionDetailViewController alloc] init];
        [detailVC setGroup:group];
        [self.parentViewController setHidesBottomBarWhenPushed:YES];
        [self.parentViewController.navigationController pushViewController:detailVC animated:YES];
        [tableView deselectRowAtIndexPath:indexPath animated:NO];
    }
}
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath{
    if (indexPath.section == 0) {
        return self.bannerData.count > 0 ? 140 : 80;
    }else if (indexPath.section == 1) {
        return 80;
    }
    return 0;
}
//MARK: TLExpressionBannerCellDelegate
- (void)expressionBannerCellDidSelectBanner:(id)item{
    WXExpressionDetailViewController *detailVC = [[WXExpressionDetailViewController alloc] init];
    [detailVC setGroup:item];
    [self.parentViewController setHidesBottomBarWhenPushed:YES];
    [self.parentViewController.navigationController pushViewController:detailVC animated:YES];
}
//MARK: TLExpressionCellDelegate
- (void)expressionCellDownloadButtonDown:(TLEmojiGroup *)group{
    group.status = TLEmojiGroupStatusDownloading;
    [self.proxy requestExpressionGroupDetailByGroupID:group.groupID pageIndex:1 success:^(id data) {
        group.data = data;
        [[WXExpressionHelper sharedHelper] downloadExpressionsWithGroupInfo:group progress:^(CGFloat progress) {
            
        } success:^(TLEmojiGroup *group) {
            group.status = TLEmojiGroupStatusDownloaded;
            NSInteger index = [self.data indexOfObject:group];
            if (index < self.data.count) {
                [self.tableView reloadRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:index inSection:1]] withRowAnimation:UITableViewRowAnimationNone];
            }
            BOOL ok = [[WXExpressionHelper sharedHelper] addExpressionGroup:group];
            if (!ok) {
                [SVProgressHUD showErrorWithStatus:[NSString stringWithFormat:@"表情 %@ 存储失败！", group.groupName]];
            }
        } failure:^(TLEmojiGroup *group, NSString *error) {
            
        }];
    } failure:^(NSString *error) {
        [SVProgressHUD showErrorWithStatus:[NSString stringWithFormat:@"\"%@\" 下载失败: %@", group.groupName, error]];
    }];
}
@end

File: Freedom/微信/我/ExpressionViewController/WXExpressionChosenViewController.swift
//
//  WXExpressionChosenViewController.swift
//  Freedom
import SnapKit
import MJRefresh
import XCarryOn
import XExtension
import Foundation
protocol WXExpressionCellDelegate: NSObjectProtocol {
    func expressionCellDownloadButtonDown(_ group: TLEmojiGroup)
}
protocol WXExpressionBannerCellDelegate: NSObjectProtocol {
    func expressionBannerCellDidSelectBanner(_ item: TLEmojiGroup)
}
class WXExpressionCell: WXTableViewCell {
    weak var delegate: WXExpressionCellDelegate?
    private lazy var iconImageView: UIImageView = {
        let iconImageView = UIImageView()
        iconImageView.backgroundColor = UIColor.clear
        iconImageView.layer.masksToBounds = true
        iconImageView.layer.cornerRadius = 5.0
        return iconImageView
    }()
    private lazy var titleLabel: UILabel = {
        let titleLabel = UILabel()
        titleLabel.font = UIFont.systemFont(ofSize: 16.0)
        return titleLabel
    }()
    private lazy var detailLabel: UILabel = {
        let detailLabel = UILabel()
        detailLabel.font = UIFont.systemFont(ofSize: 13.0)
        detailLabel.textColor = UIColor.gray
        return detailLabel
    }()
    private lazy var tagView: UIImageView = {
        let tagView = UIImageView()
        tagView.image = UIImage(named: "icon_corner_new")
        tagView.isHidden = true
        return tagView
    }()
    private lazy var downloadButton: UIButton = {
        let downloadButton = UIButton()
        downloadButton.setTitle("下载", for: .normal)
        downloadButton.titleLabel?.font = UIFont.systemFont(ofSize: 14.0)
        downloadButton.setTitleColor(UIColor.green, for: .normal)
        downloadButton.layer.masksToBounds = true
        downloadButton.layer.cornerRadius = 3.0
        downloadButton.layer.borderWidth = 1.0
        downloadButton.layer.borderColor = UIColor.green.cgColor
        downloadButton.addTarget(self, action: #selector(self.downloadButtonDown(_:)), for: .touchUpInside)
        return downloadButton
    }()
    var group: TLEmojiGroup = TLEmojiGroup() {
        didSet {
            let image = UIImage(named: group.groupIconPath)
            if image != nil {
                iconImageView.image = image
            } else {
                iconImageView.sd_setImage(with: URL(string: group.groupIconURL))
            }
            titleLabel.text = group.groupName
            detailLabel.text = group.groupDetailInfo
            if group.status == .downloaded {
                downloadButton.setTitle("已下载", for: .normal)
                downloadButton.layer.borderColor = UIColor.gray.cgColor
                downloadButton.setTitleColor(UIColor.gray, for: .normal)
            } else if group.status == .downloading {
                downloadButton.setTitle("下载中", for: .normal)
                downloadButton.layer.borderColor = UIColor.green.cgColor
                downloadButton.setTitleColor(UIColor.green, for: .normal)
            } else {
                downloadButton.setTitle("下载", for: .normal)
                downloadButton.layer.borderColor = UIColor.green.cgColor
                downloadButton.setTitleColor(UIColor.green, for: .normal)
            }
        }
    }
    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        contentView.addSubview(iconImageView)
        contentView.addSubview(titleLabel)
        contentView.addSubview(detailLabel)
        contentView.addSubview(tagView)
        contentView.addSubview(downloadButton)
        p_addMasonry()
    }
    func p_addMasonry() {
        iconImageView.snp.makeConstraints { make in
            make.left.equalTo(self.contentView).offset(15)
            make.top.equalTo(self.contentView).offset(10)
            make.bottom.equalTo(self.contentView).offset(-10)
            make.width.equalTo(self.iconImageView.height)
        }
        titleLabel.snp.makeConstraints { make in
            make.bottom.equalTo(self.iconImageView.snp.centerY).offset(-2.0)
            make.left.equalTo(self.iconImageView.right).offset(13.0)
            make.right.lessThanOrEqualTo(self.downloadButton.snp.left).offset(-15)
        }
        detailLabel.snp.makeConstraints { make in
            make.top.equalTo(self.iconImageView.snp.centerY).offset(5.0)
            make.left.equalTo(self.titleLabel)
            make.right.lessThanOrEqualTo(self.downloadButton.snp.left).offset(-15)
        }
        tagView.snp.makeConstraints { make in
            make.left.top.equalToSuperview()
        }
        downloadButton.snp.makeConstraints { make in
            make.right.equalTo(self.contentView).offset(-15)
            make.centerY.equalTo(self.contentView)
            make.size.equalTo(CGSize(width: 60, height: 26))
        }
    }
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    @objc func downloadButtonDown(_ sender: UIButton) {
        if group.status == .unDownload {
            group.status = .downloading
            delegate?.expressionCellDownloadButtonDown(group)
        }
    }
}


class WXExpressionBannerCell: WXTableViewCell,WXPictureCarouselDelegate {
    weak var delegate: WXExpressionBannerCellDelegate?
    private var picCarouselView = WXPictureCarouselView()
    var data: [WXPictureCarouselProtocol] = [] {
        didSet {
            picCarouselView.data = data
        }
    }
    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        bottomLineStyle = .none
        selectionStyle = .none
        picCarouselView.delegate = self
        contentView.addSubview(picCarouselView)
        picCarouselView.snp.makeConstraints { (make) in
            make.edges.equalTo(self.contentView)
        }
    }
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    func pictureCarouselView(_ pictureCarouselView: WXPictureCarouselView, didSelectItem model: WXPictureCarouselProtocol) {
        delegate?.expressionBannerCellDidSelectBanner(model as! TLEmojiGroup)
    }
}

class WXExpressionChosenViewController: WXTableViewController,UISearchBarDelegate {
    var kPageIndex: Int = 0
    var data: [TLEmojiGroup] = []
    var bannerData: [TLEmojiGroup] = []
    var proxy = WXExpressionHelper.shared
    private lazy var searchController: WXSearchController = {
        let searchController = WXSearchController(searchResultsController: searchVC)
        searchController.searchResultsUpdater = searchVC
        searchController.searchBar.placeholder = "搜索表情"
        searchController.searchBar.delegate = searchVC
        return searchController
    }()
    private var searchVC = WXExpressionSearchViewController()
    override func viewDidLoad() {
        super.viewDidLoad()
        automaticallyAdjustsScrollViewInsets = false
        tableView.frame = CGRect(x: 0, y: TopHeight + 20, width: APPW, height: APPH - 20 - TopHeight)
        tableView.backgroundColor = UIColor.white
        tableView.tableHeaderView = searchController.searchBar
        tableView.register(WXExpressionBannerCell.self, forCellReuseIdentifier: "TLExpressionBannerCell")
        tableView.register(WXExpressionCell.self, forCellReuseIdentifier: "TLExpressionCell")
        let footer = MJRefreshAutoNormalFooter(refreshingTarget: self, refreshingAction: #selector(self.loadMoreData))
        footer?.setTitle("正在加载...", for: .refreshing)
        footer?.setTitle("", for: .noMoreData)
        tableView.mj_footer = footer
        loadData(withLoadingView: true)
    }
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        XHud.hide()
    }
    func loadData(withLoadingView showLoadingView: Bool) {
        if showLoadingView {
            XHud.show()
        }
        kPageIndex = 1
        proxy.requestExpressionChosenList(byPageIndex: kPageIndex, success: { data in
            XHud.hide()
            self.kPageIndex += 1
            for group: TLEmojiGroup in data {
                // 优先使用本地表情
                let localEmojiGroup = WXExpressionHelper.shared.emojiGroup(byID: group.groupID)
                if let localEmojiGroup = localEmojiGroup {
                    self.data.append(localEmojiGroup)
                } else {
                    self.data.append(group)
                }
            }
            self.tableView.reloadData()
        }, failure: { error in
            XHud.hide()
        })
        proxy.requestExpressionChosenBannerSuccess({ data in
            self.bannerData = data
            self.tableView.reloadData()
        }, failure: { error in
        })
    }

    @objc func loadMoreData() {
        proxy.requestExpressionChosenList(byPageIndex: kPageIndex, success: { data in
            XHud.hide()
            if data.count == 0 {
                self.tableView.mj_footer.endRefreshingWithNoMoreData()
            } else {
                self.tableView.mj_footer.endRefreshing()
                self.kPageIndex += 1
                for group: TLEmojiGroup in data {
                    // 优先使用本地表情
                    let localEmojiGroup = WXExpressionHelper.shared.emojiGroup(byID: group.groupID)
                    if let localEmojiGroup = localEmojiGroup {
                        self.data.append(localEmojiGroup)
                    } else {
                        self.data.append(group)
                    }
                }
                self.tableView.reloadData()
            }
        }, failure: { error in
            self.tableView.mj_footer.endRefreshingWithNoMoreData()
            XHud.hide()
        })
    }
}
extension WXExpressionChosenViewController: WXExpressionCellDelegate, WXExpressionBannerCellDelegate {
    override func numberOfSections(in tableView: UITableView) -> Int {
        return bannerData.count > 0 ? 2 : 1
    }
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if section == 0 {
            return bannerData.count > 0 ? 1 : data.count
        } else if section == 1 {
            return data.count
        }
        return 0
    }
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if indexPath.section == 0 && bannerData.count > 0 {
            let bannerCell = tableView.dequeueReusableCell(withIdentifier: "TLExpressionBannerCell") as! WXExpressionBannerCell
            bannerCell.data = bannerData
            bannerCell.delegate = self
            return bannerCell
        }
        let cell = tableView.dequeueReusableCell(withIdentifier: "TLExpressionCell") as! WXExpressionCell
        let group: TLEmojiGroup = data[indexPath.row]
        cell.group = group
        cell.delegate = self
        return cell
    }
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if (indexPath.section == 0 && bannerData.count == 0) || indexPath.section == 1 {
            let group = data[indexPath.row]
            let detailVC = WXExpressionDetailViewController()
            detailVC.group = group
            parent?.hidesBottomBarWhenPushed = true
            parent?.navigationController?.pushViewController(detailVC, animated: true)
            tableView.deselectRow(at: indexPath, animated: false)
        }
    }
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        if indexPath.section == 0 {
            return bannerData.count > 0 ? 140 : 80
        } else if indexPath.section == 1 {
            return 80
        }
        return 0
    }
    //MARK: TLExpressionBannerCellDelegate
    func expressionBannerCellDidSelectBanner(_ item: TLEmojiGroup) {
        let detailVC = WXExpressionDetailViewController()
        detailVC.group = item
        parent?.hidesBottomBarWhenPushed = true
        parent?.navigationController?.pushViewController(detailVC, animated: true)
    }
    func expressionCellDownloadButtonDown(_ group: TLEmojiGroup) {
        group.status = .downloading
        proxy.requestExpressionGroupDetail(byGroupID: group.groupID, pageIndex: 1, success: { data in
            group.data = data
            WXExpressionHelper.shared.downloadExpressions(withGroupInfo: group, progress: { progress in
            }, success: { group in
                group.status = .downloaded
                let index = self.data.index(of: group) ?? -1
                if index < self.data.count {
                    self.tableView.reloadRows(at: [IndexPath(row: index, section: 1)], with: .none)
                }
                let ok = WXExpressionHelper.shared.addExpressionGroup(group)
                if !ok {
                    self.noticeError("表情 \(group.groupName) 存储失败！")
                }
            }, failure: { group, error in

            })
        }, failure: { error in
            self.noticeError("\(group.groupName)\" 下载失败: \(error)")
        })
    }
}
